### PR TITLE
feat(publish): local staged-publish pipeline with approve gate + socket scan

### PR DESCRIPTION
**Pull request by jdalton**
Adds a **local, `npm run`-driven staged-publish pipeline** to Perry. Today publishing is CI-only (`release-packages.yml`) with no approve gate and no pre-publish tarball scan; this gives the author a stage → verify → socket-scan → **human approve gate** → promote + cut-release flow run from the laptop, with OIDC trusted publishing preserved for the upload itself.

Perry's root is **npm-managed** (`package.json` + `package-lock.json`, documented in `scripts/soak/paths.mts`), so the pipeline uses **`npm stage`** (npm CLI's staged-publishing client, which ships `npm stage publish`/`approve`/`list`) — no pnpm anywhere, and `@socketsecurity/sdk` lands in the tracked `package-lock.json`.

## What lands

**`scripts/publish/` tree** (generic core + npm/brew/cargo tiers):

| Script | What it does |
|---|---|
| `npm run publish:stage` | Dispatch the new `npm-stage-publish.yml` CI workflow (OIDC staged upload: build legs reused from `release-packages.yml` stage mode → `stage-npm.sh` → `npm stage publish --provenance` for all 9 packages), then locally verify (`npm pack` sha1 vs staged shasum) + Socket-scan. **Nothing is public.** |
| `npm run publish:approve` | The promote step, run **separately** (never in CI): browser web-OTP 2FA `npm stage approve` → registry-liveness confirm → tag `vX.Y.Z` + immutable GitHub release (draft→upload `install.sh`+`checksums.txt`→undraft). |
| `npm run publish:scan` / `:status` / `:release` | Re-scan / print receipts / cut the release only. |
| `npm run publish:brew` / `:ffi` | Locally-runnable Homebrew tap bump (`PerryTS/homebrew-perry`) + perry-ffi → crates.io. |

- **`scan.mts`** — Socket full-scan of each staged tarball via `@socketsecurity/sdk` `createOrgFullScanFromArchive`; `error`-action alerts (per the org's own security policy) fail the gate, `warn`-action alerts pass with counts. Fail-closed on unreachable/empty scans.
- **`auth-posture.mts`** — refuses any long-lived `NPM_TOKEN`/`NODE_AUTH_TOKEN`/`NPM_AUTH_TOKEN` on publish (OIDC-in-CI + 2FA-locally only); `prepublishOnly` is wired to it. A read-only `PERRY_NPM_READONLY_TOKEN` powers registry reads and can never publish.
- **`release.mts`** uploads `packaging/install.sh` + `checksums.txt` as per-tag release assets, so `curl -fsSL …/releases/download/vX.Y.Z/install.sh | sh` works per tag.

**New `.github/workflows/npm-stage-publish.yml`**: OIDC staged upload (`environment: npm-publish`, `id-token: write`, no long-lived token). Stages ONLY — nothing is public until the local `publish:approve`. The existing `release-packages.yml` `npm-publish` job stays as the republish/emergency fallback (untouched).

**npm floor enforced** — `>= 11.17.0` (the newest of: `npm stage` ≥11.15.0, OIDC ≥11.5.1, `min-release-age` DAYS ≥11.17). `checkNpmFloor()` runs at the start of `pipeline.mts` (clear error + fix on older npm); the CI workflow asserts the floor in shell.

## Why stage/approve is split

The stage upload uses an OIDC token from CI; the approve requires human 2FA. Combining them would leak the OTP into CI logs or require a human at the CI keyboard — a hard requirement of npm staged publishing, not a stylistic choice.

<details><summary><b>Prerequisites the author provisions (one-time, documented in-script)</b></summary>

1. npm staged publishing enrolled for `@perryts/*` (`npm stage publish`/`approve`).
2. The new workflow added as a trusted publisher on each `@perryts/*` package (same setup `npm/README.md` describes for `release-packages.yml`).
3. `SOCKET_API_TOKEN` for tarball scans; `HOMEBREW_TAP_TOKEN`/`APT_REPO_TOKEN` for tap pushes; `PERRY_NPM_READONLY_TOKEN` for registry reads.

</details>

<details><summary><b>Verification</b></summary>

- `npm run test:scripts` — 8+ unit tests pass (formula renderer, Socket policy bucketing, human-gate block shape, auth-posture refusal sabotage-tested, `compareSemver` ordering, npm-floor coverage). Full suite 49 pass / 0 fail.
- `node scripts/publish/pipeline.mts --status` runs the import chain + reads the version gate (`v0.5.1510`); `--stage-only` dispatch reaches the workflow (404s only because the file isn't on `main` yet — expected pre-merge).
- `scripts/check_file_size.sh` green; largest new file is 276 lines.
- No version bump (external contributor); changelog fragment at `changelog.d/8336-staged-publish-pipeline.md`.

</details>

Open questions for the maintainer: (1) is `@perryts/*` enrolled in npm staged publishing (`npm stage`)? (2) the new workflow must be added as a trusted publisher on each `@perryts/*` package. The root stays npm-managed — `@socketsecurity/sdk` is synced into the tracked `package-lock.json`, and the stray root `pnpm-lock.yaml` is not used.

## Summary by CodeRabbit

* **New Features**
  * Added staged package publishing with approval-based promotion and dry-run support.
  * Added checksum verification, security scanning, registry checks, and release status tracking.
  * Added Homebrew and FFI publishing support.
  * Added automated releases with version tags, checksums, and downloadable assets.
  * Added manually triggered staging through GitHub Actions using npm trusted publishing.
* **Security**
  * Added safeguards against long-lived credentials and provenance failures.
  * Added human approval and two-factor authentication gates.
* **Documentation**
  * Documented publishing credentials and staged release procedures.
* **Tests**
  * Added coverage for publishing safeguards, approval gates, scans, and release metadata.

### CODE REVIEW

**Comment by jdalton** [comment]
Update: the pipeline now **enforces the npm floor** needed for all its features, and is fully npm (no pnpm).

**npm floor = >= 11.17.0** (the newest of the features the flow needs, so one number covers all three):
- `npm stage` (staged publishing) → >= 11.15.0
- OIDC trusted publishing → >= 11.5.1
- `min-release-age` in DAYS (the root `.npmrc` soak) → >= 11.17

**Enforced in two places:**
- **Local** — `checkNpmFloor()` runs at the start of `pipeline.mts` before any stage/approve/release action, so an author on npm <11.17 gets a clear error (with the `npm install -g npm@latest` fix) instead of a cryptic `npm stage` failure.
- **CI** — `npm-stage-publish.yml` installs `npm@latest` then asserts the floor in shell (fail-fast if a future runner image or pin ever drops below).

Since Perry's root is npm-managed, the whole flow uses `npm stage` (npm CLI ships the same staged-publishing client that pnpm also exposes) — no pnpm anywhere, and `@socketsecurity/sdk` lives in the tracked root `package-lock.json`. 49/49 script tests pass (new: `compareSemver` ordering + floor covers staged-publishing/min-release-age/OIDC).

File: .github/workflows/npm-stage-publish.yml [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            RUN_ID=$(gh run list --workflow release-packages.yml -R "$REPO" --limit 1 --json databaseId -q '.[0].databaseId' 2>/dev/null || true)
+            if [ -n "$RUN_ID" ]; then break; fi
+            sleep 3
+          done
+          if [ -z "$RUN_ID" ]; then
+            echo "::error::could not resolve a release-packages.yml run id after dispatch." >&2
+            exit 1
+          fi
+          echo "build-run-id=$RUN_ID" >> "$GITHUB_OUTPUT"
+          echo "Dispatched release-packages.yml stage build: run $RUN_ID."
```
[the remark this answers is not in the document]
Comment (jdalton) [reply .github/workflows/npm-stage-publish.yml]: Fixed in 6fb046fe4. The resolve-build job now records a dispatch timestamp, matches the workflow_dispatch run by event + createdAt (not just the newest run), and awaits it with `gh run watch --exit-status` before stage-upload downloads artifacts. Stage mode is release-packages.yml's default workflow_dispatch (no input needed), so no stage input was missing — the real defects were the newest-run selection and the missing await.

File: .github/workflows/npm-stage-publish.yml [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+  # The OIDC stage upload. Downloads the build artifacts from the resolved
+  # run, stages the npm packages, and `pnpm stage publish`es each under OIDC.
+  stage-upload:
+    needs: resolve-build
+    runs-on: ubuntu-latest
+    environment: npm-publish   # npm's trusted-publisher config pins this env name
+    permissions:
+      contents: read
+      id-token: write   # npm provenance / trusted publishing mints the OIDC token here
```
[the remark this answers is not in the document]
Comment (jdalton) [reply .github/workflows/npm-stage-publish.yml:121]: Fixed. Added `actions: read` to the stage-upload job permissions (a job-level block replaces the workflow-level one, so the workflow-level actions scope did not apply). Also moved `actions: write` off the workflow level and onto the resolve-build job only, which cleared the zizmor excessive-permissions finding.

File: .github/workflows/npm-stage-publish.yml [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            echo "PUBLISH=false (dry-run) — running stage-npm only, no registry writes."
+            exit 0
+          fi
+          # Refuse any long-lived token (the auth-posture gate, in shell form).
+          for v in NPM_TOKEN NODE_AUTH_TOKEN NPM_AUTH_TOKEN; do
+            if [ -n "$(printenv $v 2>/dev/null || true)" ]; then
+              echo "::error::Refusing to stage in CI with a long-lived $$v token — OIDC is the only sanctioned path." >&2
+              exit 1
+            fi
+          done
```
[the remark this answers is not in the document]
Comment (jdalton) [reply .github/workflows/npm-stage-publish.yml:193]: Fixed. `\$\` expands to the shell PID; changed to `\` so the error names the actual variable.

File: scripts/publish/brew/tap-publish.mts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      linuxSourceSha256 = await downloadAndSha256(sourceArchive, path.join(tmp, 'source.tar.gz'))
+    } catch (e) {
+      logger.warn(`source archive download failed (${String(e)}); linux block will carry a placeholder sha256.`)
+    }
+    const spec: PerryFormulaSpec = {
+      version,
+      macosArm64Sha256,
+      macosX64Sha256,
+      linuxSourceSha256: linuxSourceSha256 || 'PLACEHOLDER',
+    }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/brew/tap-publish.mts]: Fixed. Removed the PLACEHOLDER fallback — a failed Linux source download now fails the formula build with a clear message before the tap is cloned/pushed, since Homebrew would reject a mismatched sha256 anyway.

File: scripts/publish/npm/approve.mts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+export async function approveEntry(
+  entry: StagedEntry,
+  opts: ApproveOptions,
+): Promise<boolean> {
+  const args = ['stage', 'approve', entry.stageId]
+  if (opts.otp) args.push('--otp', opts.otp)
+  const env = { ...NON_INTERACTIVE_RENDER_ENV }
+  // --yes with no --otp → let pnpm's registry challenge drive the browser
+  // web-OTP flow. The PTY wrapper keeps the challenge alive off a non-TTY.
+  const code = await runInheritTty('pnpm', args, process.cwd(), env)
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/npm/approve.mts]: Not taking this one. The code invokes `npm` (not `pnpm`); `PNPM_CONFIG_OTP` is a pnpm v11 env var that npm does not read, so the proposed fix would break the OTP flow. `--otp` is npm's documented non-interactive path. The argv-exposure concern is low: the code is a ~30s-expiring TOTP run on a local terminal, and the CI path uses the browser web-OTP flow (--yes), not --otp.

File: scripts/publish/npm/approve.mts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  // 4. pnpm stage approve each entry.
+  const approvedEntries: StagedEntry[] = []
+  const failed: string[] = []
+  for (const entry of verified) {
+    const ok = await approveEntry(entry, opts)
+    ;(ok ? approvedEntries : failed).push(entry)
+  }
+  const approved = approvedEntries.map(e => `${e.name}@${e.version}`)
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/npm/approve.mts:189]: Fixed. `failed.push(entry)` was pushing a StagedEntry into a `string[]`; now pushes `${entry.name}@${entry.version}` to match the ApproveReceipt.failed contract every other return path satisfies.

File: scripts/publish/npm/staged.mts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+      continue
+    }
+    const { version } = readPackageJson(pkgDir)
+    logger.log(`=== direct publish ${name}@${version ?? '?'} ===`)
+    const result = await uploadNpmPackage({ cwd: pkgDir, dryRun, mode: 'direct', tag })
+    if (!result.postureOk || result.code !== 0) {
+      failed.push(`${name}@${version}`)
+      continue
+    }
+    published.push(`${name}@${version}`)
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/npm/staged.mts:153]: Fixed. runDirect now guards a missing version (matching runStaged) so it no longer attempts a publish for a manifest with no version or records `name@undefined`.

File: scripts/publish/pipeline.mts [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+  logger.log(`  approved:  ${state.approved?.length ? state.approved.join(', ') : '(not approved)'}`)
+  logger.log(`  registry:  ${state.registryLive ? 'live' : 'not live'}`)
+  logger.log(`  released:  ${state.released ? 'yes' : 'no'}`)
+}
+
+async function main(): Promise<void> {
+  const flags = new Set(process.argv.slice(2))
+  const dryRun = flags.has('--dry-run')
+  const noScan = flags.has('--no-scan')
+  const tag = (flags.has('--tag') && process.argv[process.argv.indexOf('--tag') + 1]) || 'latest'
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/pipeline.mts]: Fixed. `--tag` and `--otp` values are parsed out of argv before the flag set is built, so `--tag beta` no longer adds `beta` to the flag set and break mode routing. A value equal to a flag name can't mis-route either (it's consumed as the option's value).

File: scripts/publish/publish.test.mts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  const orig = process.env['NPM_TOKEN']
+  process.env['NPM_TOKEN'] = 'npm_secret_live_token'
+  try {
+    const reason = publishAuthPreflight({ ci: true, direct: false, dryRun: true })
+    assert.equal(reason, undefined, 'dry-run must skip the posture gate')
+  } finally {
+    if (orig === undefined) delete process.env['NPM_TOKEN']
+    else process.env['NPM_TOKEN'] = orig
+  }
+})
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/publish.test.mts]: Fixed. Tests now snapshot, clear, and restore all three long-lived token vars (NPM_TOKEN, NODE_AUTH_TOKEN, NPM_AUTH_TOKEN), and add refusal cases for each plus the 0.0.0 reservation escape hatch.

File: scripts/publish/release.mts [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    )
+    if (remote.code !== 0 || !remote.stdout.includes(`refs/tags/${tagName}`)) {
+      logger.fail(
+        `could not push tag ${tagName} to origin (git push exited ${pushed.code}) and origin does not carry it.\n` +
+          `  Fix: resolve the push (auth? protected ref? network?) and re-run.`,
+      )
+      process.exitCode = 1
+      return false
+    }
+    logger.log(`Tag ${tagName} already on origin; continuing.`)
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/release.mts]: Fixed. On a failed push, the remote tag's peeled commit is compared to the local tag's commit; a mismatch (someone else tagged a different commit) fails loud instead of continuing to publish against the wrong commit.

File: scripts/publish/release.mts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  // 3. If the release already exists, leave it untouched.
+  const view = await runCapture('gh', ['release', 'view', tagName, '--json', 'tagName'], rootPath)
+  if (view.code === 0) {
+    logger.log(`Release ${tagName} already exists; leaving it untouched.`)
+    return true
+  }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/release.mts]: Fixed. An existing release is now inspected for `isDraft`; a draft is resumed (asset upload + undraft) rather than treated as complete. A published release is left untouched.

File: scripts/publish/release.mts [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+  // 5. Assets: install.sh + checksums.txt.
+  const installSh = path.join(rootPath, INSTALL_SH)
+  const assets: string[] = []
+  if (existsSync(installSh)) {
+    const checksums = writeChecksums([installSh])
+    assets.push(installSh, checksums)
+  } else {
+    logger.warn(`install.sh not found at ${installSh} — releasing without it.`)
+  }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/release.mts]: Fixed. install.sh is now required BEFORE the tag is created or pushed — a missing install.sh fails before any immutable release marker exists, rather than warning after the tag is live.

File: scripts/publish/scan.mts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+  try {
+    await sdk.getQuota()
+  } catch (e) {
+    logger.fail(
+      `Socket auth failed: getQuota() rejected the token (${String(e)}).\n` +
+        `  Fix: check ${SOCKET_TOKEN_ENV_VAR} at ${SOCKET_TOKEN_MINT_URL}.`,
+    )
+    return undefined
+  }
+  return { orgSlug, sdk }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/scan.mts:154]: Fixed. `getQuota().success` is now checked; an unsuccessful response (e.g. a 401 on an invalid token that doesn't throw) is treated as auth failure and fails closed.

File: scripts/publish/scan.mts [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+  ctx: SocketScanContext,
+  name: string,
+  version: string,
+): Promise<ScanResult> {
+  const { orgSlug, sdk } = ctx
+  const pkgDir = path.join(rootPath, npmPackageDir(name))
+  const packed = await packTarball(pkgDir)
+  if (!packed || !existsSync(packed.path)) {
+    logger.fail(`scan: no local tarball for ${name}@${version} — run verify first.`)
+    return { name, version, status: 'blocked', summary: { error: [], total: 0, warn: [] } }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/scan.mts:186]: Fixed. scanTarball now accepts the staged entry's shasum and verifies the freshly-packed tarball matches it before submitting, so a worktree change between verify and scan can't make Socket scan different bytes from the artifact promotion releases.

File: scripts/publish/shared.mts [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+    }
+    if (inString) continue
+    if (ch === '{') depth += 1
+    else if (ch === '}') {
+      depth -= 1
+      if (depth === 0) return text.slice(startIdx, i + 1)
+    }
+  }
+  return undefined
+}
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/shared.mts:285]: Fixed. extractFirstJson now honors `[` as a start token and matches the balanced `]`, so array JSON from `npm stage list --json` parses correctly instead of extracting only the first inner object. Regression tests added for clean arrays, noisy-wrapped arrays, and brace-in-string edge cases.

File: scripts/publish/auth-posture.mts [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  if (!direct) {
+    return (
+      `Refusing to stage with a long-lived token present: ${present.join(', ')}. ` +
+      'Staging runs in CI under OIDC; locally only the 2FA approve is sanctioned. ' +
+      'Unset the token and dispatch the stage workflow instead.'
+    )
+  }
+  // direct local publish: permitted only for the 0.0.0 reservation escape hatch.
+  return undefined
+}
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/auth-posture.mts:89]: Fixed. publishAuthPreflight now takes a version; a direct publish with a long-lived token is refused except for 0.0.0. uploadNpmPackage reads the manifest version and passes it, so the escape hatch can't be widened into a general bypass.

File: scripts/publish/npm/approve.mts [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+      scanResults.push(res)
+      if (res.status === 'passed') passed.push(entry)
+      else logger.fail(`scan ${res.status}: ${entry.name}@${entry.version} dropped from approve`)
+    }
+    if (passed.length === 0) {
+      logger.fail('No staged entries passed the scan gate — not approving.')
+      return { approved: [], failed: verified.map(e => `${e.name}@${e.version}`), registryLive: false, scanResults }
+    }
+    verified.length = 0
+    verified.push(...passed)
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/npm/approve.mts:168]: Fixed. runApprove now requires every ALL_PACKAGES member to be staged, and rejects at the verify and scan gates if any expected package is missing or fails — a partial promote of the fixed 9-package set can no longer ship a broken install.

File: scripts/publish/auth-posture.mts [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+  direct?: boolean
+  dryRun?: boolean
+  version?: string
+}): string | undefined {
+  const { ci = isCI(), direct = false, dryRun = false, version } = config
+  if (dryRun) return undefined
+  const present = LONG_LIVED_NPM_TOKEN_ENV_VARS.filter(
+    v => process.env[v] !== undefined && process.env[v] !== '',
+  )
+  if (present.length === 0) return undefined
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/auth-posture.mts:64]: Not taking this one in this PR. The auth-posture gate is env-var-based by design — it mirrors the CI OIDC path, which sets no .npmrc tokens. Expanding it to scan .npmrc configs (project root, user config dir, NPM_CONFIG_USERCONFIG, scoped registries) with correct read-only-token discrimination is a separate hardening pass that needs its own tested PR; a naive text scan would false-positive on the read-only PERRY_NPM_READONLY_TOKEN. The direct escape hatch is now restricted to 0.0.0, and CI (the primary publish path) uses OIDC with no .npmrc tokens. Tracking as a follow-up.

File: scripts/publish/brew/tap-publish.mts [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    const clone = await runCapture(
+      'git',
+      ['clone', `https://x-access-token:${token}@github.com/${HOMEBREW_TAP_REPO}.git`, tapDir],
+      os.tmpdir(),
+    )
+    if (clone.code !== 0) {
+      logger.fail(`git clone of ${HOMEBREW_TAP_REPO} failed (${clone.code}).`)
+      process.exitCode = 1
+      return
+    }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/brew/tap-publish.mts:99]: Fixed in cdd90f093. The clone now uses the plain HTTPS URL with a git credential helper that reads HOMEBREW_TAP_TOKEN from the (inherited) environment, so the token stays out of the process argument list and out of .git/config. The push uses the same helper.

File: scripts/publish/brew/tap-publish.mts [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+      [['commit', '-m', commitMsg], 'git commit'],
+      [['push', 'origin', 'HEAD'], 'git push'],
+    ] as [string[], string][]) {
+      const r = await runCapture('git', args, tapDir)
+      if (r.code !== 0) {
+        logger.fail(`${label} failed (${r.code}).`)
+        process.exitCode = 1
+        return
+      }
+    }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/brew/tap-publish.mts:130]: Fixed in cdd90f093. After git add, a `git diff --cached --quiet` check skips the commit + push when the formula is unchanged (the tap already carries it for that version) instead of failing on git commit's nothing-to-commit exit.

File: scripts/publish/npm/approve.mts [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+export async function approveEntry(
+  entry: StagedEntry,
+  opts: ApproveOptions,
+): Promise<boolean> {
+  const args = ['stage', 'approve', entry.stageId]
+  if (opts.otp) args.push('--otp', opts.otp)
+  const env = { ...NON_INTERACTIVE_RENDER_ENV }
+  // --yes with no --otp → let npm's registry challenge drive the browser
+  // web-OTP flow. The PTY wrapper keeps the challenge alive off a non-TTY.
+  const code = await runInheritTty('npm', args, process.cwd(), env)
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/npm/approve.mts:71]: Fixed in cdd90f093. The OTP is now passed through NPM_CONFIG_OTP (npm's `npm_config_*` env channel) instead of `--otp <code>`, so the ~30s-expiring code is not visible in the process argument list. npm reads this env var at the enforced floor (>= 11.17). Your re-flag correctly identified NPM_CONFIG_OTP for npm — the original suggestion (PNPM_CONFIG_OTP) was pnpm-specific and would not have worked here.

File: scripts/publish/npm/approve.mts [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+
+  // 5. Registry liveness — exit 0 is NOT proof.
+  let registryLive = true
+  for (const entry of approvedEntries) {
+    if (!(await fetchPublishedVersion(entry.name, entry.version))) {
+      logger.fail(`registry liveness: ${entry.name}@${entry.version} not resolvable after approve — do NOT cut the release.`)
+      registryLive = false
+    }
+  }
+  return { approved, failed, registryLive, scanResults }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/npm/approve.mts:209]: Fixed in cdd90f093. registryLive now starts false when failed.length > 0, and a loud message reports the partial promote. This closes the last gap in the no-partial-set contract (verify, scan, and approve all refuse a subset now).

File: scripts/publish/release.mts [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+        logger.log(`Release ${tagName} already published; leaving it untouched.`)
+        return true
+      }
+      logger.log(`Release ${tagName} exists as a draft — resuming asset upload + publish.`)
+      const upload = await runCapture('gh', ['release', 'upload', tagName, ...assets], rootPath)
+      if (upload.code !== 0) {
+        logger.fail(`gh release upload failed (${upload.code})`)
+        process.exitCode = 1
+        return false
+      }
```
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/release.mts]: Fixed in cdd90f093. Added --clobber to the draft-resume upload (and the initial create upload for consistency) so an interrupted run's already-uploaded install.sh / checksums.txt are overwritten instead of failing the retry. checksums.txt is regenerated each run, so overwriting is the intended behavior.

File: scripts/publish/auth-posture.mts (hunk 18 above)
[the remark this answers is not in the document]
Comment (jdalton) [reply scripts/publish/auth-posture.mts:64]: The auth-posture gate deliberately uses browser web-OTP 2FA locally + OIDC in CI, not the `PNPM_CONFIG_OTP` env var — a long-lived OTP in the env is the kind of token the gate exists to refuse. The other 7 findings on this PR (dispatch/await exact run, actions: read, $$v, checksum fail-closed, failed type, version guard, flag-set pollution) are all addressed in the current commit (cdd90f093) — the review comments are against earlier commits.
